stress/v2/ponyExpress/util.go

```diff
@@ -0,0 +1,88 @@
+package ponyExpress
+
+import (
+	"crypto/rand"
+	"fmt"
+	"strconv"
+	"sync"
+)
+
+// ###########################################
+// ConcurrencyLimiter and associated methods #
+// ###########################################
+
+// ConcurrencyLimiter ensures that no more than a specified
+// max number of goroutines are running.
+type ConcurrencyLimiter struct {
```

Comment:
This seems like a very complicated way to limit the size of a pool of goroutines. Perhaps I'm missing something, but if you're basically trying to do a simple (non-leaky) bucket then you can achieve this with a simple buffered channel acting as a semaphore. Something like:

```go
type ConcurrencyLimiter struct {
	sem              chan struct{}
}

func NewConcurrencyLimiter(max int) *SemaphoreLimiter {
	return &SemaphoreLimiter{
		sem: make(chan struct{}, max),
	}
}

func (c *ConcurrencyLimiter) Increment() {
	c.sem <- struct{}{}
}

func (c *ConcurrencyLimiter) Decrement() {
	// So we don't block if the number of goroutines is currently 0.
	select {
	case <-c.sem:
	default:
	}
}
```